File: old-projects/vector-defence-classic-port/src/Particles/Projectile.ts
import { Point } from "../Point";
import { Sprite } from "../Sprite";
import { Game } from "../Game";
import { Util } from "../Util";

export class Projectile extends Sprite {
  static readonly speed = 7;
  private readonly damage: number;

  constructor(source: Point, dest: Point, damageCaused: number, size: number) {
    super();
    this.damage = damageCaused;
    this.x = source.x;
    this.y = source.y;
    const angle = Util.calculateAngle(this.x, this.y, dest.x, dest.y);
    const velocity = Util.calculateLocation(angle, Projectile.speed);
    this.dx = velocity.x;
    this.dy = velocity.y;
    this.radius = size / 2;
  }

  override animate(): void {
    this.updatePosition();
    if (this.isOutsideBounds) {
      this.markedForRemoval = true;
      return;
    }

    for (const monster of Game.currentFrameMonsters) {
      if (Util.isWithinDistance(monster.x, monster.y, this.x, this.y, monster.radius + this.radius)) {
        monster.takeDamage(this.damage);
        this.markedForRemoval = true;
        return;
      }
    }
  }

  override render(context: CanvasRenderingContext2D): void {
    context.save();
    context.fillStyle = "#9fffe4";
    context.beginPath();
    context.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
    context.fill();
    context.restore();
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Particles/Projectile.cs
using System.Windows;
using System.Windows.Media;
using System.Windows.Shapes;

namespace VectorDefenceSL.Particles
{
    public class Projectile : Sprite
    {
        public const double Speed = 7;
        
        private readonly double damage;
        private static readonly SolidColorBrush ProjectileFill = Util.GetColorFromHex("9FFFE4");

        public Projectile(Point source, Point dest, double damageCaused, double size)
        {
            damage = damageCaused;
            X = source.X;
            Y = source.Y;

            double angle = Util.CalculateAngle(X, Y, dest.X, dest.Y);
            Util.CalculateLocation(angle, Speed, out DX, out DY);

            mainElement = new Ellipse
            {
                Fill = ProjectileFill,
                Width = size,
                Height = size,
                IsHitTestVisible = false,
            };
            Radius = size / 2;
            Game.AddElement(mainElement);      
        }

        public override void Animate()
        {
            UpdatePosition();

            if (IsOutsideBounds)
            {
                MarkedForRemoval = true;
                return;
            }

            foreach (var m in Game.CurrentFrameMonsters)
            {
                bool closeEnough = Util.IsWithinDistance(m.X, m.Y, X, Y, m.Radius + Radius);
                if (closeEnough)
                {
                    m.TakeDamage(damage);
                    MarkedForRemoval = true; // projectile should be removed
                    return;
                }
            }
        }

        public override void Render()
        {
            Move(X - Radius, Y - Radius);
        }
    }
}
